controllers/apim/ratelimitpolicy_controller.go

```diff
@@ -85,15 +100,343 @@ func (r *RateLimitPolicyReconciler) Reconcile(eventCtx context.Context, req ctrl
 }
 
+func (r *RateLimitPolicyReconciler) reconcileWithVirtualService(ctx context.Context, vs *istio.VirtualService, rlp *apimv1alpha1.RateLimitPolicy) error {
+	logger := r.Logger()
+
+	_, ok := vs.Annotations[VSAnnotationRateLimitPolicy]
+	if !ok {
+		rlpNamespacedName := client.ObjectKeyFromObject(rlp)
+		vs.Annotations[VSAnnotationRateLimitPolicy] = rlpNamespacedName.String()
+		if err := r.Client().Update(ctx, vs); err != nil {
+			logger.Error(err, "failed to add ratelimitpolicy annotation to virtualservice")
+			return err
+		}
+		logger.V(1).Info("successfully added ratelimitpolicy annotation to virtualservice")
+	}
+
+	// TODO(rahulanand16nov): store context of virtualservice in RLP's status block and manage envoy patches.
+	if err := r.reconcileRateLimitFilters(ctx, vs, rlp); err != nil {
+		logger.Error(err, "failed to reconcile ratelimit filters")
+		return err
+	}
+
+	if err := r.reconcileRateLimitActions(ctx, vs, rlp); err != nil {
+		return err
+	}
+	logger.Info("successfully reconciled ratelimitpolicy using attached virtualservice")
+	return nil
+}
+
+func (r *RateLimitPolicyReconciler) reconcileRateLimitActions(ctx context.Context, vs *istio.VirtualService, rlp *apimv1alpha1.RateLimitPolicy) error {
+	logger := r.Logger()
+
+	// TODO(rahulanand16nov): right now patch only applies to kuadrant-gateway but ideally should choose from virtualservice.
+	for _, host := range vs.Spec.Hosts {
+		for _, route := range rlp.Spec.Routes {
+			isRouteMatch := doesRouteMatch(vs.Spec.Http, route.Name)
+			if !isRouteMatch {
+				logger.V(1).Info("no match found for route", "route name", route.Name)
+				continue
+			}
+
+			vHostName := host + ":80" // Istio names virtual host in this format: host + port
+			routePatch := routeRateLimitsPatch(vHostName, route.Name, route.Actions, route.Stage)
+
+			if err := controllerutil.SetOwnerReference(rlp, routePatch, r.Client().Scheme()); err != nil {
+				logger.Error(err, "failed to add owner ref to route-level ratelimits patch")
+				return err
+			}
+			if err := r.Client().Create(ctx, routePatch); err != nil {
+				if !apierrors.IsAlreadyExists(err) {
+					logger.Error(err, "failed to create route specific rate limit patch %s", routePatch.Name)
+					return err
+				}
+			}
+			logger.Info("created route level ratelimits patch", "name", routePatch.Name)
+		}
+	}
+	return nil
+}
+
+func (r *RateLimitPolicyReconciler) reconcileRateLimitFilters(ctx context.Context, vs *istio.VirtualService, rlp *apimv1alpha1.RateLimitPolicy) error {
+	logger := r.Logger()
+
+	for _, gw := range vs.Spec.Gateways {
+		gwKey := gatewayNameToObjectKey(gw, vs.Namespace) // Istio defaults to vs namespace
+		rateLimitPatch := rateLimitInitialPatch(gwKey)
+
+		if err := controllerutil.SetOwnerReference(rlp, rateLimitPatch, r.Client().Scheme()); err != nil {
+			logger.Error(err, "failed to add owner ref to RateLimit filters patch")
+			return err
+		}
+		if err := r.Client().Create(ctx, rateLimitPatch); err != nil {
```

Comment:
Objects needs to be reconciled, not only created. Otherwise changes in the RLP will not update the envoyfilter objects.

Anyway, it is good enough for now

Reply:
Yes, just waiting for the green light before i tighten those loose screws.

Reconciling EnvoyFilter is a bit tricky since need to be aware of VS/HTTPR change as well.